### ISSUE TITLE
warp terminates silently

### ISSUE THREAD

**Issue by agrafix** [yesodweb/wai#603]
On a system I've observed that a web server application written using `warp` sometimes terminates with `exit code 0` for no reason. After some debugging, I noticed that it ran out of file descriptors. The problem lies here: https://github.com/yesodweb/wai/blob/c6a542cc4bc51106b207432b1ef81e8dd675a323/warp/Network/Wai/Handler/Warp/Run.hs#L224-L253 . With the default configuration, the exception is NOT logged and the runLoop simply terminates. I think we should always log the expection in default configuration if the server terminates.

**Comment by snoyberg** [comment]
Reading through that code, shouldn't `settingsOnException` be able to print the exception?

**Comment by agrafix** [comment]
Yes, but I think the default `defaultOnException` with `defaultShouldDisplayException` misses the out of file descriptors exceptions and thus does not print it

**Comment by snoyberg** [comment]
Here's the relevant code:

https://www.stackage.org/haddock/lts-7.14/warp-3.2.9/src/Network.Wai.Handler.Warp.Settings.html#defaultShouldDisplayException

Which of those clauses is catching the out of file descriptors exception?

**Comment by Vlix** [comment]
I think this issue has been solved by catching the FD exhaustion (which has been in the code for a while now) and either waiting for the amount of connections to go down, or to just try waiting for another connection (after #1084)
